### PR TITLE
Refactor mini-tools

### LINKED ISSUE #133: Update mini-tools to be more standardized to the "Rails way" of building interactive elements

### PR DESCRIPTION
**Pull request by josefarias**
closes https://github.com/maybe-finance/marketing/issues/133

This PR refactors the marketing site's mini-tools to follow Hotwire and Rails conventions.

Previously, the tools used a bespoke template rendering system instead of Hotwire (Turbo)'s built-in frames. Leveraging Hotwire makes the code:

- Cleaner, because we don't need to manually parse form data, or manually render a template, or define that template. Instead, we can just submit the form and let Turbo do the heavy lifting.
- OSS-friendly, because it now looks like any other Hotwire codebase. Contributors will recognize patterns from the Hotwire docs and other Hotwire codebases.
- More Rails-like, since we're now rendering plain old erb (e.g. `<%= tool.inflation_rate %>` instead of `<span t-text="inflationRate"></span>`)

Also, all calculations were previously done on the client. They're now happening on the back-end. This makes the code:

- Simpler, because there are less moving parts since we don't need to pass data to the client for processing.
- Easier to read, because of Ruby's expressiveness and pithy syntax.
- Easier to maintain, as I've taken the opportunity to remove some indirection.
- Testable (which, this PR adds tests for all tools) — testing server code with unit tests is easier than emulating a headless client with system tests.

Some JS cleanup was done in passing. But I decided not to make it a priority because the scope is already large enough and the existing code is not significantly breaking convention. There's room for improvement, but it's less crucial than the other problems addressed by this PR. That means there will still be work to be done on JS after merging. Most notably, the time series charts have not yet been refactored to be reusable — they remain coupled to specific tools.

Finally, I've removed code comments — which seem to have been added recently, and intentionally. I'm happy to bring those back, but I want to make the case that code comments add a maintainability burden in keeping them up to date. Ideally, the code should be readable enough that comments are mostly not necessary.

> [!IMPORTANT]
> This touches every single mini-tool. We should do some manual QA on every tool to ensure consistency.

Steps to add a new tool (for future reference)
---
1. Add tool to `db/seeds/tools.rb`
2. Create the tool record:
    1. Local: Run `bin/rails db:seed`
    2. Prod: Create the tool by the usual means, without running the seeds. I'm unaware whether the usual means is through Avo, or directly in the console. But the seed data doesn't match what's on prod today.
4. Create a presenter under `app/presenters/tool`.
    1. Class name must match the tool's slug.
    3. All tool presenters must implement the `#active_record` method to find their corresponding tool in the DB.
    4. See neighboring presenters for reference.
5. Create a partial under `app/views/tools/widgets`
    1. Partial name must match the tool's slug.
    2. Currently, all tools are further broken down in `app/views/tools/widgets/forms` and `app/views/tools/widgets/content`. It's a nice pattern, but not mandatory.
    3. See neighboring partials for reference.
6. Allow-list the presenter's params in `ToolsController#tool_params`.
7. You're set! Use ruby to implement business logic in the presenter. Render as usual:
    1. `<%= tool.some_financial_indicator %>`
    2. `<%= number_with_delimiter(tool.some_financial_indicator) %>`
    5. `<%= number_to_currency(tool.some_financial_indicator) %>`

### CODE REVIEW

**Comment by zachgoll** [comment]
🚀 🚀

File: app/controllers/tools_controller.rb
Comment (josefarias) [inline app/controllers/tools_controller.rb:23]: This might be too clever. I might redirect instead and create one controller for each tool. It's more code, but ultimately the most idiomatic and flexible approach.

File: app/presenters/tool/bogleheads/position.rb
Comment (josefarias) [inline app/presenters/tool/bogleheads/position.rb:41]: Should probably store as a hash grouped by date and find the stock price that way instead of finding iteratively

Comment (josefarias) [reply app/presenters/tool/bogleheads/position.rb:41]: I'll also add a db index for the ticker

File: config/environments/production.rb
Comment (josefarias) [inline config/environments/production.rb:62]: Heads up — this'll need to be configured before deploying.

We're already using `Rails.cache.fetch`. But it's not doing anything because no cache store is configured. This'll fix that.

File: db/migrate/20240915055133_add_index_for_stock_price_tickers.rb
Comment (josefarias) [inline db/migrate/20240915055133_add_index_for_stock_price_tickers.rb:3]: Heads up — this migration will need to be ran after deploying. We're already fetching by ticker, and we're leaning into that even more in this PR. This'll make the queries quicker. Not that they'd absolutely need it now, but at some point they will.

File: app/presenters/tool/presenter.rb
Comment (josefarias) [inline app/presenters/tool/presenter.rb:21]: Possibly a job for `ActiveModel::Attributes`

File: app/javascript/controllers/time_series_bar_chart_controller.js
Comment (josefarias) [inline app/javascript/controllers/time_series_bar_chart_controller.js:36]: Fixes duplicate charts when navigating back and forth using the browser history

File: app/helpers/tools_helper.rb
Comment (josefarias) [inline app/helpers/tools_helper.rb:4]: Heads up — removing this guard means we'll try (and fail) to render tools that exist in the db but don't have an associated partial.

I think ideally we shouldn't have tools like that. And I think we likely don't. But thought it was worth calling out.

File: app/models/tool.rb
Comment (josefarias) [inline app/models/tool.rb:55]: Can skip this

**Comment by josefarias** [comment]
Practically done here, just not comfortable marking it as ready because the stock portfolio backtest combobox search stopped working for me locally for some reason. It's also happening on `main`, but I want to understand what's going on there before merging.

Gonna be attending Rails World for the next couple of days, so I most likely won't be making progress until I'm back. But this is more than ready for feedback @zachgoll.

**Comment by zachgoll** [comment]
@josefarias awesome, I'll take a read-through of this after I get my CSV imports overhaul done, which should line up pretty well with your return from Rails world!  Have a good time, hoping to make it there one of these years :)

**Review by boowjoow — APPROVED** [review]

**Comment by josefarias** [comment]
@zachgoll I've figured out the search problem. Nothing to worry about, just local env trouble.

Tests are passing locally but CI was failing this weekend. I think I've managed to address that but we'll need another CI run to make sure.

Marking as ready now.

Before deploying, I'd recommend:
1. Configuring Redis for the Rails cache, this should be an isolated DB not shared with other concerns like background jobs (mentioned in a comment)
2. Running the migration in this PR (mentioned in a comment)
3. Deploying to a staging environment and giving the tools a quick manual test, since this touches a ton of code it's entirely possible we missed something. I don't expect anything to be broken though, as I've done manual testing myself and also added unit tests for all tools.
4. Making sure the server is not under-provisioned. Probably a non-issue but previously we might've gotten away with a super tiny box. We're gonna be performing calculations there now, so possibly worth a quick sense check.

File: config/application.rb
Comment (josefarias) [inline config/application.rb:26]: Stops deprecation warnings when running tests.

File: config/environments/production.rb
Comment (zachgoll) [reply config/environments/production.rb:62]: Thanks for the heads up.  cc @Shpigford, we switched this over to Hatchbox recently correct?

**Comment by zachgoll** [comment]
@josefarias thanks for the helpful context and summary, looking through this now!

File: app/presenters/tool/presenter.rb
Comment (zachgoll) [inline app/presenters/tool/presenter.rb:11]: Is there a reason we're storing the tools in the database (and seeding them) rather than just hardcoding the information into the presenter and pointing a new route at them?

Mainly just trying to understand the pattern as I'm just getting up to speed with this codebase.  The current workflow to create a new tool would be this correct?

1. Write migration to add a tool (assuming migration rather than seed since we only want to run the seed once in prod?)
2. Create presenter

File: config/initializers/tool_attributes.rb
Comment (zachgoll) [inline config/initializers/tool_attributes.rb:73]: Nice, wouldn't of thought to do this!

File: app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb
Comment (zachgoll) [inline app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb]: Probably worth throwing a `required: true` on this one since it raises if not populated

Also, as i was testing this tool locally, I could not get it to spit out any results.  Loaded stocks to the DB, but there might be other required data I'm missing?

https://github.com/user-attachments/assets/78c89115-5e47-4488-afbe-f7b1ab2f59cb

**Review by zachgoll — COMMENTED** [review]
Overall, love the changes!  Looked through the models / controllers / views pattern and it was super easy to follow as someone who hasn't looked at this codebase much, so hopefully the same holds true for new contributors :)

Left a few comments/questions, but I wouldn't let any of them hold this up.  I did some manual QA and everything looked great.

File: app/presenters/tool/presenter.rb
Comment (josefarias) [reply app/presenters/tool/presenter.rb:11]: > Is there a reason we're storing the tools in the database (and seeding them) rather than just hardcoding the information into the presenter and pointing a new route at them?

I don't see a compelling reason, just that it's the way it was built initially.

I also had the instinct to hardcode the info. I'd be +1 to refactor things that way at some point.

Persisting in the DB is the reason this comment was necessary: https://github.com/maybe-finance/marketing/pull/134#discussion_r1760164050

> The current workflow to create a new tool would be this correct?

I've updated the PR description with the steps to add a new tool. The seeds in this repo seem to have been thought out to populate the dev env, but never the prod env. Another thing I'd be +1 to change.

File: app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb
Comment (josefarias) [reply app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb]: @zachgoll good catch! I've added `required: true` to both dates.

> Loaded stocks to the DB, but there might be other required data I'm missing?

Ah, it's because the start and end dates need to be on different months, since we query Synth on a monthly interval to get the stock prices. Pushed this up to make that a bit more clear: https://github.com/maybe-finance/marketing/pull/134/commits/c44535e1430520444c16e570db35a84f6cd4f60f

Although the text is slightly incorrect. They don't need to be a month apart, just on different months. But struggled to find good wording for that.

File: app/presenters/tool/presenter.rb
Comment (zachgoll) [reply app/presenters/tool/presenter.rb:11]: RE: storing in DB—Gotcha, makes sense.  No need to change, just wanted to make sure I was thinking about that correctly for future reference.

RE: seeds—really appreciate you adding that "how to" in the PR description.

I know this is a bit "non-standard", but given we've got all the tools in the DB and need parity between dev/prod, what if the instructions added a sample migration for creating a new tool?  That way, a contributor could simply write a migration that could be run in both dev or prod for the tool?

```rb
class AddSomeNewTool < ActiveRecord::Migration[8.0]
  def up
    Tool.create!(
      name: "Stock Price Calculator",
      slug: "stock-price-calculator",
      intro: "Calculate and analyze stock prices with ease.",
      description: "Our Stock Price Calculator is a powerful tool designed to help investors and financial analysts quickly compute and evaluate stock prices.",
      content: "Use this tool to:\n\n- Calculate current stock values\n- Analyze historical price trends\n- Compare multiple stocks\n- Estimate future stock prices based on various metrics",
      category_slug: "finance",
      icon: "calculator",
      meta_image_url: "https://example.com/images/stock-calculator.jpg"
    )
  end

  def down
    Tool.find_by(slug: "stock-price-calculator")&.destroy
  end
end
```

File: app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb
Comment (zachgoll) [reply app/views/tools/widgets/forms/_stock_portfolio_backtest.html.erb]: Got it, makes sense!  I think the wording update works!

File: app/presenters/tool/presenter.rb
Comment (josefarias) [reply app/presenters/tool/presenter.rb:11]: I'm not strongly opposed to that, but I think that'd be over-indexing on a pattern we'd actively like to get rid of.

We don't currently have parity between dev and prod, so instead of introducing a migration pattern to have parity going forward, I think ideally we'd fix at the root and just do away with db-backed tools in one fell swoop. Specially because I don't think it'd be a heavy lift.

BTW I just thought of a reason we might want the tools on the db. And that's if Avo provides WYSIWYG editing for the text. I'm not sure it does.

In the end, I'd make this decision based on who's gonna be editing the contents. Probably not the OSS devs, right? So that's another reason doing the migrations might not be as helpful — we wouldn't be able to merge + deploy without proofreading/editing.

As far as db-backed vs hardcoded, I'd say it depends on whether the person editing the contents is comfortable writing markdown in a heredoc, or whether they'd be more productive using WYSIWYG.

Comment (zachgoll) [reply app/presenters/tool/presenter.rb:11]: Ahhh I think you might be right about Avo (I had not seen that when I was going through and reviewing).

Let's keep it as you've got it, I agree with the rationale here and you're almost certainly correct in saying that OSS contributors wouldn't be editing the _existence_ of tools anyways.

**Comment by zachgoll** [comment]
@josefarias everything looks good here to me.  Will let @Shpigford do the merging and bounty related stuff!

Thanks again for the awesome work here!

**Comment by josefarias** [comment]
Always a pleasure @zachgoll, @Shpigford! I'll be available outside of work to troubleshoot any regressions, should they come up.

Just pushed up some minor nit fixes, this is ready to merge.

**Pull request merged**
